### ISSUE TITLE
"Very slow compile" of dm-haiku training step

### ISSUE THREAD

**Issue by nalzok** [google/jax#10790]
Please:

- [X] Check for duplicate issues.
- [X] Provide a complete example of how to reproduce the bug, wrapped in triple backticks like this:

```python
import itertools

import jax
import jax.numpy as jnp

import haiku as hk
import optax


Model = hk.nets.ResNet152


def dataloader(batch_size, *, key):
    while True:
        # Binary classification
        X_key, Y_key = jax.random.split(key)
        X = jax.random.normal(X_key, (batch_size, 64, 64, 3))
        Y = jax.random.randint(Y_key, (batch_size,), 0, 2)
        yield X, Y


def train():
    batch_size = 32
    learning_rate = 0.001

    def forward(input, is_training):
        net = Model(num_classes=2)
        output = net(input, is_training)
        return output

    def learner_fn(input, ground_truth):
        logits = forward(input, True)
        labels = jax.nn.one_hot(ground_truth, 2)
        return jnp.mean(optax.softmax_cross_entropy(logits, labels))

    learner_fn_t = hk.transform_with_state(learner_fn)
    learner_fn_t = hk.without_apply_rng(learner_fn_t)

    @jax.jit
    def train_step(params, state, opt_state, X, y):
        (loss, state), grads = jax.value_and_grad(learner_fn_t.apply, has_aux=True)(params, state, X, y)
        updates, opt_state = optimizer.update(grads, opt_state, params)
        params = optax.apply_updates(params, updates)
        return params, state, opt_state, loss
    
    rng = jax.random.PRNGKey(42)
    loader_rng, rng = jax.random.split(rng)
    loader = dataloader(batch_size, key=loader_rng)

    X, y = next(loader)
    params, state = learner_fn_t.init(rng, X, y)

    optimizer = optax.adam(learning_rate)
    opt_state = optimizer.init(params)

    # Training
    for X, y in itertools.islice(loader, 50):
        params, state, opt_state, loss = train_step(params, state, opt_state, X, y)
        print(loss)


if __name__ == '__main__':
    train()
```

- [X] If applicable, include full error messages/tracebacks.

```
$ XLA_FLAGS=--xla_dump_to=/tmp/slow_operation_dump pipenv run python3 slow_operation.py
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
2022-05-21 01:56:39.688109: E external/org_tensorflow/tensorflow/compiler/xla/service/slow_operation_alarm.cc:61]
********************************
Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
Compiling module jit_train_step.261
********************************
```

Here is the content of `/tmp/slow_operation_dump`: [slow_operation_dump.zip](https://github.com/google/jax/files/8747119/slow_operation_dump.zip). Additionally, here is the Pipenv lock files to help you reproduce my development environment on macOS 12.3.1: [Pipfile.zip](https://github.com/google/jax/files/8747124/Pipfile.zip)

**Comment by nalzok** [comment]
Okay, after waiting for ~10 minutes the JIT compilation appears to be successful, and the loss is dropping rapidly! However, the backpropagation is very slow (each iteration takes several seconds), so I suspect there are still some problems.

```python
$ XLA_FLAGS=--xla_dump_to=/tmp/slow_operation_dump pipenv run python3 slow_operation.py
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
2022-05-21 01:56:39.688109: E external/org_tensorflow/tensorflow/compiler/xla/service/slow_operation_alarm.cc:61]
********************************
Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
Compiling module jit_train_step.261
********************************
0.6931475
0.657604
0.4421207
0.19778039
0.046985622
0.01008308
0.002680283
0.00088729034
0.00035099505
0.00015979768
8.119534e-05
4.4985038e-05
2.6750724e-05
1.6849244e-05
1.1138508e-05
7.696395e-06
5.513401e-06
4.071727e-06
3.0957071e-06
2.4139822e-06
1.9259712e-06
1.5795204e-06
1.3075752e-06
1.1026847e-06
9.499482e-07
8.232885e-07
7.338816e-07
6.4820006e-07
5.8114495e-07
5.3271623e-07
4.8801274e-07
4.507599e-07
4.1350705e-07
3.9115534e-07
3.6135302e-07
3.4645188e-07
3.2037485e-07
3.1292427e-07
2.9802314e-07
2.8312198e-07
2.7567143e-07
2.6822084e-07
2.6449555e-07
2.5704497e-07
2.5331968e-07
2.421438e-07
2.3096796e-07
2.3096796e-07
2.2351738e-07
2.1979208e-07
```

Three additional files appeared in `/tmp/slow_operation_dump`, but somehow I cannot upload them to GitHub directly. Let me know if they are important, and I can set up a Google Drive link for you.

```
$ diff -rq /tmp/slow_operation_dump /tmp/slow_operation_dump_old
Only in /tmp/slow_operation_dump/: module_0282.jit_train_step.261.ir-with-opt-noconst.ll
Only in /tmp/slow_operation_dump/: module_0282.jit_train_step.261.ir-with-opt.ll
Only in /tmp/slow_operation_dump/: module_0282.jit_train_step.261.o
```

**Comment by tomhennigan** [comment]
`ResNet152` is a fairly compute intensive model (forward pass is ~12GFLOPs) so I'm not surprised that the step time to train it in full precision on CPU is in the order of a second or two. You would probably be better of training this on a GPU or TPU (or using a smaller ResNet model like ResNet50)?

The XLA:GPU and XLA:TPU compilers are also better optimized (vs. XLA:CPU) and so compilation will also likely be quicker.

Taking your code and running it on a GPU Colab instance it compiles in ~2 mins and the step time (including printing the loss each step) is 200ms. If you make the loss printing less frequent your step time would likely be a lot faster.

https://colab.research.google.com/gist/tomhennigan/79ecdf274ad4c36a21a64cd4a18a4042/google-jax-10790.ipynb

Note that I did refactor your code a little, to try and ensure JAX operations were wrapped into three big jits (one to generate dataset batches, one to generate all initial state, and the other to compute a training step).

**Comment by nalzok** [comment]
Thanks for the reply, @tomhennigan. I wonder why we should JIT the `initial_state` function, though. This function is called once, so I'm not sure if it's worth the compilation overhead.

**Comment by tomhennigan** [comment]
The way JAX works is that when you execute a single op eagerly (e.g. `a + b`) JAX will just in time compile a program to add two tensors of the given shape/dtype. So If you don't `jit` your `initial_state` function JAX will actually just in time compile hundreds of small kernels anyway, one for each op.

I would suspect that compilation of 100s of small programs vs. compilation of one program with 100s of ops would take a similar amount of time. The larger program may be a bit slower to compile (there are some compiler passes that scale non-linearly in the number of ops in the program) but I would guess the timings would be pretty close.

Your thinking is definitely correct, that it is far more important to `jit` the `train_step` function.

There are two other reasons why jit compiling initial_state might be preferable:

1. **Fragmentation** - if you don't jit compile initial_state, then JAX needs to allocate and de-allocate every tensor in your initial_state program eagerly. If you imagine memory as a list of values, this means your parameters and temporaries would be allocated like so: `[param1, temp,, .., param2, temp, ..]`. When the `initial_state` function exits all the temporaries will be deallocated, so you are left with fragmentation: `[param1, <free>, param2, <free>, ...]`. If you JIT compile, JAX allocates space for all outputs first, followed by space for all temporaries, so your memory should be less fragmented: `[param1, param2, <free>, ..]`. (The details of memory allocation may vary between JAX backends and I am not an expert here, but this is my understanding of how it typically works. Fragmentation also might not be something you care about if you have plenty of GPU memory available to execute your programs, so you may only want to do this optimisation when you need it.) 

3. **Future upgrades** - We find that JAX users want to use >1 GPU, and the simplest way today to do that is `pmap` and there you would need to `pmap` your `initial_state` function (to replicate your parameters across multiple GPUs).

**Comment by nalzok** [comment]
Thank you, @tomhennigan. That is extremely helpful!
